obkvrpc/connection_pool.go

```diff
@@ -139,11 +178,223 @@ func (p *ConnectionPool) CreateConnection(ctx context.Context) (*Connection, err
+	}
+	slbAddress := strset.NewWithSize(len(ips))
+	for _, ip := range ips {
+		slbAddress.Add(ip.String())
+	}
+	changed := !slbAddress.IsEqual(strset.New(s.slbAddress...))
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	if changed {
+		log.Info(fmt.Sprint("SLB address changed, before: ", s.slbAddress, ", after: ", slbAddress))
+		s.slbAddress = slbAddress.List()
+	}
+	return changed, nil
+}
+
+// round-robin get next slb address from slb list
+func (s *SLBLoader) getNextSLBAddress() string {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+	slbNum := len(s.slbAddress)
+	if slbNum > 0 {
+		slbAddr := s.slbAddress[(s.round.Add(1))%(int64(slbNum))]
+		return slbAddr
+	}
+	return s.dnsAddress
+}
+
+// refresh SLBList and refresh connection expire time if SLBList changed
+func (s *SLBLoader) run() {
+	changed, err := s.refreshSLBList()
+	if err != nil {
+		log.Warn("reconnect failed", zap.Error(err))
+		return
+	}
+	if changed {
+		s.refreshConnectionLife()
+	}
+}
+
+func (s *SLBLoader) refreshConnectionLife() {
+	for _, conn := range s.connPool.connections {
+		conn.expireTime = time.Now()
+	}
+}
+
+func (s *SLBLoader) toString() string {
+	return fmt.Sprintf("%#v", s)
+}
+
+// check and reconnect timeout connections
+func (c *ConnectionLifeCycleMgr) run() {
+	if c.connPool == nil {
+		log.Error("connection pool is null")
+		return
+	}
+
+	// 1. get all timeout connections
+	expiredConnIds := make([]int, 0, len(c.connPool.connections))
+	//for idx, connection := range c.connPool.connections {
+	for i := 1; i <= len(c.connPool.connections); i++ {
+		connection := c.connPool.connections[(i+c.lastExpireIdx)%(len(c.connPool.connections))]
+		if !connection.expireTime.IsZero() && connection.expireTime.Before(time.Now()) {
+			expiredConnIds = append(expiredConnIds, (i+c.lastExpireIdx)%(len(c.connPool.connections)))
+		}
+	}
+
+	if len(expiredConnIds) > 0 {
+		log.Info(fmt.Sprintf("Find %d expired connections", len(expiredConnIds)))
+		for idx, connIdx := range expiredConnIds {
+			log.Info(fmt.Sprintf("%d: ip=%s, port=%d", idx, c.connPool.connections[connIdx].option.ip, c.connPool.connections[connIdx].option.port))
+		}
+	}
+
+	// 2. mark 30% expired connections as expired
+	maxReconnIdx := int(math.Ceil(float64(len(expiredConnIds)) / 3))
+	if maxReconnIdx > 0 {
+		c.lastExpireIdx = expiredConnIds[maxReconnIdx-1]
+		log.Info(fmt.Sprintf("Begin to refresh expired connections which idx less than %d", maxReconnIdx))
+	}
+	for i := 0; i < maxReconnIdx; i++ {
+		// no one can get expired connection
+		c.connPool.connections[expiredConnIds[i]].isExpired.Store(true)
+	}
+	defer func() {
+		for i := 0; i < maxReconnIdx; i++ {
+			c.connPool.connections[expiredConnIds[i]].isExpired.Store(false)
+		}
+	}()
+
+	// 3. wait all expired connection finished
+	time.Sleep(DefaultConnectWaitTime)
+	for i := 0; i < maxReconnIdx; i++ {
+		pool := c.connPool.connections
+		idx := expiredConnIds[i]
+		for j := 0; len(pool[idx].pending) > 0; j++ {
+			if j > 0 && j%10000 == 0 {
```

Comment:
It seems that you are using a special way to wait for the pending to get into 0, why do we wait like this?

Reply:
It's wait for the pending finish in busy way, and print the log in per-10000 cycle, after discuss, change the busy wait way to sleep to avoid using to much resource